src/wms/src/main/java/org/geoserver/wms/RenderingVariablesCallback.java

```diff
@@ -0,0 +1,108 @@
+    public static final String WMS_HEIGHT = "wms_height";
+    /** The GetMap scale denominator, as an {@link Double} */
+    public static final String WMS_SCALE_DENOMINATOR = "wms_scale_denominator";
+
+    public static final String WMS_CRS = "wms_crs";
+
+    public static void setupEnvironmentVariables(WMSMapContent mapContent) {
+        // setup some SLD variable substitution environment used by rendering transformations
+        EnvFunction.setLocalValue(WMS_BBOX, mapContent.getRenderingArea());
+        EnvFunction.setLocalValue(WMS_CRS, mapContent.getRenderingArea().getCoordinateReferenceSystem());
+        EnvFunction.setLocalValue(WMS_SRS, mapContent.getRequest().getSRS());
+        EnvFunction.setLocalValue(WMS_WIDTH, mapContent.getMapWidth());
+        EnvFunction.setLocalValue(WMS_HEIGHT, mapContent.getMapHeight());
+        try {
+            double scaleDenominator = mapContent.getScaleDenominator(true);
+            EnvFunction.setLocalValue(WMS_SCALE_DENOMINATOR, scaleDenominator);
+        } catch (Exception e) {
+            LOGGER.log(
+                    Level.SEVERE,
+                    "Failed to compute the scale denominator, wms_scale_denominator env variable is unset",
+                    e);
+        }
+    }
+
+    @Override
+    public GetMapRequest initRequest(GetMapRequest request) {
+        return request;
+    }
+
+    @Override
+    public void initMapContent(WMSMapContent mapContent) {}
+
+    @Override
+    public Layer beforeLayer(WMSMapContent mapContent, Layer layer) {
+        if (layer instanceof FeatureLayer) {
+            @SuppressWarnings("unchecked")
+            Map<String, Object> virtualTableParamsMap =
+                    (Map<String, Object>) layer.getQuery().getHints().get(VIRTUAL_TABLE_PARAMETERS);
+            if (virtualTableParamsMap == null) {
+                virtualTableParamsMap = new HashMap<>();
+                layer.getQuery().getHints().put(VIRTUAL_TABLE_PARAMETERS, virtualTableParamsMap);
+            }
+
+            Map<String, Object> localValues = EnvFunction.getLocalValues();
+            setValueIfNotPresent(virtualTableParamsMap, localValues, WMS_BBOX);
+            setValueIfNotPresent(virtualTableParamsMap, localValues, WMS_CRS);
+            setValueIfNotPresent(virtualTableParamsMap, localValues, WMS_SRS);
+            setValueIfNotPresent(virtualTableParamsMap, localValues, WMS_WIDTH);
+            setValueIfNotPresent(virtualTableParamsMap, localValues, WMS_HEIGHT);
+            setValueIfNotPresent(virtualTableParamsMap, localValues, WMS_SCALE_DENOMINATOR);
+        }
+
+        return layer;
+    }
+
+    private static void setValueIfNotPresent(
+            Map<String, Object> virtualTableParamsMap, Map<String, Object> localValues, String key) {
+        if (!virtualTableParamsMap.containsKey(key) && localValues.containsKey(key.toUpperCase())) {
+            virtualTableParamsMap.put(key, localValues.get(key.toUpperCase()));
```

Comment:
It seems you haven't actually tested this change interactively... unless I'm missing something.
See the code expanding the variables down in GeoTools:

https://github.com/geotools/geotools/blob/afdeb4f33495eab8bbbc2ee0785222cfda881b4e/modules/library/jdbc/src/main/java/org/geotools/jdbc/VirtualTable.java#L190

The code expects a Map<String, String>, virtual table parameters just does a string expansion. With this code, I would expect to see a class cast exception when it does:

```java
            String value = values.get(param.getName());
```